File: packages/relay-core/src/relay-error.ts
/**
 * {@link RelayError} — a typed failure carrying the cell's {@link RelayCloseCode}.
 *
 * When a cell refuses a connection (an unknown host, a bad or expired ticket, a
 * failed proof, a bridge timeout, a draining cell) it sends a `close { code }`
 * outer message. The adapters surface that as a rejected promise whose error is a
 * `RelayError` carrying the `code`, so a caller can branch on *why* it was
 * refused rather than parsing a string.
 */
import type { RelayCloseCode } from './messages.js'

/** A relay refusal, carrying the coded reason the cell reported. */
export class RelayError extends Error {
  /** The close code the cell sent. */
  readonly code: RelayCloseCode
  /** The optional human-readable reason, if the cell supplied one. */
  readonly reason: string | undefined

  constructor(code: RelayCloseCode, reason?: string) {
    super(reason === undefined ? `relay closed: ${code}` : `relay closed: ${code} (${reason})`)
    this.name = 'RelayError'
    this.code = code
    this.reason = reason
  }
}

File: ios/PherryKit/Sources/PherryKit/RelayError.swift
import Foundation

/// A relay refusal, carrying the coded reason a cell reported.
///
/// When a cell refuses a data connection (an unknown host, a bad / expired / reused ticket, a
/// draining cell, a bridge timeout) it sends a `close { code }` outer message; the dialer
/// surfaces that as a thrown `RelayError` whose `code` a caller can branch on. `code` is `nil`
/// only for a local protocol fault (a malformed or unexpected outer frame) with no cell code.
public struct RelayError: Error, Sendable, Equatable {
    /// The cell's close code (e.g. `bad-ticket`, `drained`), or `nil` for a local protocol fault.
    public let code: String?
    /// A human-readable description; never carries a secret.
    public let message: String

    /// Build a relay error from an optional code and a message.
    public init(code: String?, message: String) {
        self.code = code
        self.message = message
    }
}
